src/ZODB/DB.py

```diff
@@ -1056,19 +1058,38 @@ class TransactionalUndo(object):
 
     def __init__(self, db, tids):
         self._db = db
-        self._storage = getattr(
-            db._mvcc_storage, 'undo_instance', db._mvcc_storage.new_instance)()
         self._tids = tids
+        self._storage = None
 
     def abort(self, transaction):
         pass
 
+    def close(self):
+        if self._storage is not None:
+            self._storage.close()
+            self._storage = None
+
     def tpc_begin(self, transaction):
+        assert self._storage is None, "Already in an active transaction"
```

Comment:
Should we raise a proper exception here?

Reply:
I think a failure of that condition would mean that the `transaction` library/protocol is broken. So to test it we'd have to use this object in a very unnatural way, violating its contract. ZODB contains numerous `assert` statements for other conditions that "can't" happen. I didn't see the value in making this one different and writing the corresponding tests.